patches/server/0918-Add-damage-source-wrapper.patch

```diff
@@ -0,0 +1,826 @@
+                 }
+-                event = new EntityDamageByEntityEvent(damager.getBukkitEntity(), entity.getBukkitEntity(), damageCause, modifiers, modifierFunctions, source.isCritical()); // Paper - add critical damage API
++                event = new EntityDamageByEntityEvent(damager.getBukkitEntity(), entity.getBukkitEntity(), damageCause, origin, modifiers, modifierFunctions, source.isCritical()); // Paper - add critical damage API
+             }
+             event.setCancelled(cancelled);
+ 
+-            CraftEventFactory.callEvent(event);
+-
+-            if (!event.isCancelled()) {
++            if (event.callEvent()) { // Paper
+                 event.getEntity().setLastDamageCause(event);
+             } else {
+                 entity.lastDamageCancelled = true; // SPIGOT-5339, SPIGOT-6252, SPIGOT-6777: Keep track if the event was canceled
+@@ -1008,9 +1006,9 @@ public class CraftEventFactory {
+                 cause = DamageCause.SONIC_BOOM;
+             }
+ 
+-            return CraftEventFactory.callEntityDamageEvent(damager, entity, cause, modifiers, modifierFunctions, cancelled, source.isCritical()); // Paper - add critical damage API
++            return CraftEventFactory.callEntityDamageEvent(damager, entity, cause, origin, modifiers, modifierFunctions, cancelled, source.isCritical()); // Paper - add critical damage API // Paper - Damage source wrapper
+         } else if (source == DamageSource.OUT_OF_WORLD) {
+-            EntityDamageEvent event = new EntityDamageByBlockEvent(null, entity.getBukkitEntity(), DamageCause.VOID, modifiers, modifierFunctions);
++            EntityDamageEvent event = new EntityDamageByBlockEvent(null, entity.getBukkitEntity(), DamageCause.VOID, origin, modifiers, modifierFunctions); // Paper - Damage source wrapper
+             event.setCancelled(cancelled);
+             CraftEventFactory.callEvent(event);
+             if (!event.isCancelled()) {
+@@ -1020,7 +1018,7 @@ public class CraftEventFactory {
+             }
+             return event;
+         } else if (source == DamageSource.LAVA) {
+-            EntityDamageEvent event = (new EntityDamageByBlockEvent(CraftEventFactory.blockDamage, entity.getBukkitEntity(), DamageCause.LAVA, modifiers, modifierFunctions));
++            EntityDamageEvent event = (new EntityDamageByBlockEvent(CraftEventFactory.blockDamage, entity.getBukkitEntity(), DamageCause.LAVA, origin, modifiers, modifierFunctions)); // Paper - Damage source wrapper
+             event.setCancelled(cancelled);
+ 
+             Block damager = CraftEventFactory.blockDamage;
+@@ -1045,10 +1043,8 @@ public class CraftEventFactory {
+                 cause = DamageCause.MAGIC;
+             } else if (source == DamageSource.IN_FIRE) {
+                 cause = DamageCause.FIRE;
+-            } else {
+-                throw new IllegalStateException(String.format("Unhandled damage of %s by %s from %s", entity, damager, source.msgId));
+             }
```

Comment:
Have this log a warning instead, and say that the damage cause may not be accurate and to use the damage source api instead.